// @tbd54566975/dwn-sdk-js/dist/esm/src/handlers/events-query.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { Events } from '../utils/events.js';
import { EventsQuery } from '../interfaces/events-query.js';
import { messageReplyFromError } from '../core/message-reply.js';
import { authenticate, authorizeOwner } from '../core/auth.js';
export class EventsQueryHandler {
    constructor(didResolver, eventLog) {
        this.didResolver = didResolver;
        this.eventLog = eventLog;
    }
    handle({ tenant, message }) {
        return __awaiter(this, void 0, void 0, function* () {
            let eventsQuery;
            try {
                eventsQuery = yield EventsQuery.parse(message);
            }
            catch (e) {
                return messageReplyFromError(e, 400);
            }
            try {
                yield authenticate(message.authorization, this.didResolver);
                yield authorizeOwner(tenant, eventsQuery);
            }
            catch (e) {
                return messageReplyFromError(e, 401);
            }
            const eventFilters = Events.convertFilters(message.descriptor.filters);
            const { events, cursor } = yield this.eventLog.queryEvents(tenant, eventFilters, message.descriptor.cursor);
            return {
                status: { code: 200, detail: 'OK' },
                entries: events,
                cursor
            };
        });
    }
}
