// @tbd54566975/dwn-sdk-js/dist/esm/src/interfaces/events-subscribe.js
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
import { AbstractMessage } from '../core/abstract-message.js';
import { Message } from '../core/message.js';
import { removeUndefinedProperties } from '../utils/object.js';
import { Time } from '../utils/time.js';
import { DwnInterfaceName, DwnMethodName } from '../enums/dwn-interface-method.js';
import { validateProtocolUrlNormalized, validateSchemaUrlNormalized } from '../utils/url.js';
export class EventsSubscribe extends AbstractMessage {
    static parse(message) {
        return __awaiter(this, void 0, void 0, function* () {
            Message.validateJsonSchema(message);
            yield Message.validateSignatureStructure(message.authorization.signature, message.descriptor);
            for (const filter of message.descriptor.filters) {
                if ('protocol' in filter && filter.protocol !== undefined) {
                    validateProtocolUrlNormalized(filter.protocol);
                }
                if ('schema' in filter && filter.schema !== undefined) {
                    validateSchemaUrlNormalized(filter.schema);
                }
            }
            Time.validateTimestamp(message.descriptor.messageTimestamp);
            return new EventsSubscribe(message);
        });
    }
    /**
     * Creates a EventsSubscribe message.
     *
     * @throws {DwnError} if json schema validation fails.
     */
    static create(options) {
        var _a, _b;
        return __awaiter(this, void 0, void 0, function* () {
            const currentTime = Time.getCurrentTimestamp();
            const descriptor = {
                interface: DwnInterfaceName.Events,
                method: DwnMethodName.Subscribe,
                filters: (_a = options.filters) !== null && _a !== void 0 ? _a : [],
                messageTimestamp: (_b = options.messageTimestamp) !== null && _b !== void 0 ? _b : currentTime,
            };
            removeUndefinedProperties(descriptor);
            const authorization = yield Message.createAuthorization({
                descriptor,
                signer: options.signer
            });
            const message = { descriptor, authorization };
            Message.validateJsonSchema(message);
            return new EventsSubscribe(message);
        });
    }
}
